module: add warnings for experimental flags

pd4d10 opened the pull request:
Fixes: https://github.com/nodejs/node/issues/30600

### Progress

`--es-module-specifier-resolution=node` has been added in the PR at #30678.

#### Warning Message
- [x] `--experimental-json-modules`
- [x] `--experimental-wasm-modules`
- [x] `--experimental-resolve-self`
- [x] `--experimental-conditional-exports`

#### Test Case
- [x] `--experimental-json-modules`
- [x] `--experimental-wasm-modules`
- [x] `--experimental-resolve-self`
- [x] `--experimental-conditional-exports`

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

guybedford commented on lib/internal/modules/esm/default_resolve.js:
```diff
@@ -43,11 +43,15 @@ const legacyExtensionFormatMap = {
 
-if (experimentalJsonModules)
+if (experimentalJsonModules) {
+  process.emitWarning('The JSON module is experimental.', 'ExperimentalWarning');
```
The actual usage of these is for example at https://github.com/nodejs/node/blob/master/lib/internal/modules/esm/translators.js#L136.

I think moving the warning to the actual usage could be beneficial.

guybedford commented on src/module_wrap.cc:
```diff
@@ -1143,6 +1146,7 @@ Maybe<URL> ResolveSelf(Environment* env,
     return Nothing<URL>();
   }
 
+  ProcessEmitWarning(env, "The resolve self is experimental.");
```
Again the warning should be on the return path (the only return path that isn't Nothing).

Also there is a corresponding path for this in the CJS loader.

guybedford commented on src/module_wrap.cc:
```diff
@@ -954,15 +955,17 @@ Maybe<URL> ResolveExportsTarget(Environment* env,
+              conditionalTarget, subpath, pkg_subpath, base, false);
+        if (!resolved.IsNothing()) {
+          return resolved;
+        }
```
Nice work getting this in the C++! Thought that was the major tricky one here. Note the standard warning message structure, it would be goo dot ensure we are following this. Note there is also a path for this in https://github.com/nodejs/node/blob/master/lib/internal/modules/cjs/loader.js#L577.

pd4d10 commented:
@guybedford Thanks for your review!

I've noticed that there is already a function in `lib/internal/util.js` to ensure that experimental message only show once, by saving shown flags to a set and check it at next time.

https://github.com/nodejs/node/blob/0646eda4fc0affb98e13c30acb522e63b7fd6dde/lib/internal/util.js#L168

Shall we reuse this function?

guybedford commented:
Yes that is the function to use. Is there a version of that for C++ to use
or do we have to create our own?

On Mon, Nov 25, 2019 at 09:01 Rongjian Zhang <notifications@github.com>
wrote:

> @guybedford <https://github.com/guybedford> Thanks for your review!
>
> I've noticed that there is already a function in lib/internal/util.js to
> ensure that experimental message only show once, by saving shown flags to a
> set and check it at next time.
>
>
> https://github.com/nodejs/node/blob/0646eda4fc0affb98e13c30acb522e63b7fd6dde/lib/internal/util.js#L168
>
> Shall we reuse this function?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/30617?email_source=notifications&email_token=AAESFSRMRNLUXRSRIWPZO5LQVPLE3A5CNFSM4JQ5LIBKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFCPVPQ#issuecomment-558168766>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAESFSSSUBHPZY4QTV34IJDQVPLE3ANCNFSM4JQ5LIBA>
> .
>

addaleax commented:
This LGTM code-wise :+1:

guybedford commented on src/module_wrap.cc:
```diff
@@ -7,6 +7,7 @@
 #include "util-inl.h"
 #include "node_contextify.h"
 #include "node_watchdog.h"
+#include "node_options.h"
```
Probably need to include `node_process.h`?

pd4d10 replied:
Whoops. Good catch, thanks!

Now I'm trying to figure out how to make experimental warnings emitted only once in C++. And also, add related test cases.

After all the staff done I'll `at` you to review again.

pd4d10 commented:
Hi @guybedford , I've updated the current progress at https://github.com/nodejs/node/pull/30617#issue-344872621.

There are some questions I haven't figured out and I'd like to ask you:

- `--es-module-specifier-resolution=node` doesn't have `experimental` prefix. Shall we use the same warning message as other experimental flags? The current message is:

```sh
{{Feature name}} is an experimental feature. This feature could change at any time
```

- `--experimental-resolve-self` and `--experimental-conditional-exports` was once tested in `test/es-module/test-esm-exports.mjs`, and removed in this PR: https://github.com/nodejs/node/pull/29866. Shall we continue to write test cases for these two flags in this file(`test-esm-exports.mjs`), or create new one?

Thanks!

guybedford commented:
> --es-module-specifier-resolution=node doesn't have experimental prefix. Shall we use the same warning message as other experimental flags? The current message is:

This has been added in the PR at https://github.com/nodejs/node/pull/30678. We're discussing it at https://github.com/nodejs/node/pull/30678/files#r351640229.

> Shall we continue to write test cases for these two flags in this file(test-esm-exports.mjs), or create new one?

The test-esm-exports.mjs test runs with the `--experimental-modules` flag which is currently an alias for the `--experimental-resolve-self` and `--experimental-conditional-exports` flags. So the test still includes them.

A new test for the warning messages only could be worthwhile though.

pd4d10 commented on test/common/fixtures.mjs:
```diff
@@ -0,0 +1,20 @@
+/* eslint-disable node-core/require-common-first, node-core/required-modules */
```
Add `common/fixtures.mjs` for fixtures path resolve utility in ES modules. Please comment here if there is a better way to do it

guybedford replied:
You should just be able to import `common` to do this?

pd4d10 replied:
Yeah, I've tried to find it at [common/index.mjs](https://github.com/nodejs/node/blob/master/test/common/index.mjs) at first. There seems no such method to do fixtures path resolve. Please let me know if I missed

guybedford replied:
Agreed fixtures.mjs is necessary.

guybedford commented on test/es-module/test-esm-wasm.mjs:
```diff
@@ -13,3 +15,23 @@ strictEqual(addImported(0), 42);
+  strictEqual(code, 0);
+  strictEqual(signal, null);
+  ok(stderr.toString().includes(
+    'ExperimentalWarning: WASM Module is an experimental feature. ' +
```
Perhaps - `WASM module imports are an experimental feature`, similarly for the others.

guybedford replied:
To ensure the wording here, let's make the feature name _**Importing WASM modules** is an experimental feature_

pd4d10 replied:
OK, it is more clear than before. Updated

guybedford commented:
I also believe that the JS warning methods don't automatically dedupe themselves like this C++ one does. That seems to be an argument for having boolean flags at the calling points... but I'm not set either way on this.

pd4d10 commented on src/node_process_events.cc:
```diff
@@ -95,6 +96,21 @@ Maybe<bool> ProcessEmitWarning(Environment* env, const char* fmt, ...) {
+  std::string message(warning);
+  message.append(
+      " is an experimental feature. This feature could change at any time");
+  return ProcessEmitWarningGeneric(env, message.c_str(), "ExperimentalWarning");
```
Tested this warning in C++, seems there are already `(node:49426) ExperimentalWarning:` prefix 
without any code change... But I currently don't found the related code in C++.

I need to confirm again for this point

guybedford replied:
Can we mark this as resolved?

pd4d10 commented on src/node_process_events.cc:
```diff
@@ -95,6 +96,21 @@ Maybe<bool> ProcessEmitWarning(Environment* env, const char* fmt, ...) {
+  experimental_warnings.insert(warning);
+  std::string message(warning);
+  message.append(
+      " is an experimental feature. This feature could change at any time");
```
OK. This implementation refers to this method in JS code:

https://github.com/nodejs/node/blob/b419d2823037a6662d52b47a4b4fac1aa64e5548/lib/internal/util.js#L170-L176

Perhaps we should make these two implementation unified

guybedford replied:
Ah I completely missed that the inconsistency is that we aren't using this path to warn experimental-modules itself! In that case, your implementation is the better approach and we should ignore my comments :)

guybedford commented on test/common/fixtures.mjs:
```diff
@@ -0,0 +1,20 @@
+import { createRequire } from 'module';
+
+const require = createRequire(import.meta.url);
+const fixtures = require('./fixtures.js');
```
This can also be achieved with just:

```js
export { fixturesDir, path, readSync, readKey } from './fixtures.js';
```

pd4d10 replied:
OK. Updated

guybedford commented on test/es-module/test-esm-json.mjs:
```diff
@@ -1,7 +1,29 @@
+  strictEqual(code, 0);
+  strictEqual(signal, null);
+  ok(stderr.toString().includes(
+    'ExperimentalWarning: JSON Module is an experimental feature. ' +
```
Similarly, I think we should make this _**Importing JSON modules** is an experimental feature_.

pd4d10 replied:
Updated

guybedford commented on src/module_wrap.cc:
```diff
@@ -962,6 +963,7 @@ Maybe<URL> ResolveExportsTarget(Environment* env,
       Maybe<URL> resolved = ResolveExportsTarget(env, pjson_url,
             conditionalTarget, subpath, pkg_subpath, base, false);
       if (!resolved.IsNothing()) {
+        ProcessEmitExperimentalWarning(env, "Conditional Exports");
```
We should make `exports` lowercase here.

pd4d10 replied:
Updated

guybedford commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -700,6 +704,7 @@ Module._findPath = function(request, paths, isMain) {
 
   const selfFilename = trySelf(paths, exts, isMain, trailingSlash, request);
   if (selfFilename) {
+    emitExperimentalWarning('Resolve Self');
```
For this one, perhaps _**Package name self resolution** is an experimental feature_. @jkrems does that work for you here?

jkrems replied:
Sounds good! Thanks for all your work on this, @pd4d10!

pd4d10 replied:
+1. Updated

joaocgreis commented:
Windows CI re-run: https://ci.nodejs.org/job/node-test-commit-windows-fanned/31996/

guybedford commented:
Landed in https://github.com/nodejs/node/commit/aa4c57ae7e74a623f5255a7f05a6ff9d5bc335d0.
